streams: Add admin UI for making private stream history public to members.

eeshangarg opened the pull request:
This is a part of #9232.

@timabbott: FYI, I plan on continuing this tomorrow! I labelled it as a WIP but I think we may be able to merge commits as we go! Thanks for trusting me with this project! :)

eeshangarg commented:
@timabbott: Updated! :)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1564,14 +1564,22 @@ def send_stream_creation_event(stream: Stream, user_ids: List[int]) -> None:
+            history_public_to_subscribers = False
+        else:
+            history_public_to_subscribers = True
+
```
The correct logic for this, at the current stage, is the below:

```
diff --git a/zerver/lib/actions.py b/zerver/lib/actions.py
index 04b9ecf..9dd10b6 100644
--- a/zerver/lib/actions.py
+++ b/zerver/lib/actions.py
@@ -1567,11 +1567,21 @@ def create_stream_if_needed(realm: Realm,
                             invite_only: bool=False,
                             history_public_to_subscribers: Optional[bool]=None,
                             stream_description: Text="") -> Tuple[Stream, bool]:
-    if history_public_to_subscribers is None:
-        if invite_only:
-            history_public_to_subscribers = False
-        else:
-            history_public_to_subscribers = True
+    if invite_only:
+        if history_public_to_subscribers is None:
+            # TODO: Once we have a UI for this feature, we'll remove
+            # settings.PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS and set
+            # this to be False here
+            history_public_to_subscribers = settings.PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS
+    else:
+        # If we later decide to support public streams without
+        # history, we can remove this code path.
+        history_public_to_subscribers = True
+
+    if realm.is_zephyr_mirror_realm:
+        # In the Zephyr mirroring model, history is unconditionally
+        # not public to subscribers, even for public streams.
+        history_public_to_subscribers = False
 
     (stream, created) = Stream.objects.get_or_create(
         realm=realm,
```

Three key reasons:
(1) You misread the carefully written code in models.py around `is_public`, which checks is_zephyr_mirror_realm in addition to `invite_only`, and thus that logic was being ignored here (which would have created problems for any newly created streams).
(2) Until the frontend offers an option for the new property explicitly in the UI, any users who were previously using `PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS` would have had that setting be broken for any new streams created after we merge this commit (but before we added the UI).  So fixing this is a prerequisite to be able to merge these commits now, rather than having to wait for the UI commits to all be ready.
(3) Your original implementation enabled the unintended feature of supporting making a public stream without public history.  We may perhaps want such a feature eventually, but there's no reason to offer it by accident.

I think that's the only real issue with this PR, so I went ahead and made those tweaks, added tests for the new code paths, and merged the commit so far with those changes.

timabbott commented:
I merged the first commit as 621350070241cd478610efd8f2bacf14a4127d79; see my comments above on the changes I made.

A general note here on refactoring: When making changes to existing code, especially backend code relevant to the security model, it's important to be really thoughtful about whether you're changing behavior in unintended ways, since that's the stuff of which security vulnerabilities is made :).  I hope the changes I made noted above help you get a feel for the types of issues to think about when doing this sort of important backend change (in particular, I'm pretty glad to have been able to merge this 250-line commit now, so I don't have to review it again, and that imposes some requirements on how the commit is implemented so that we don't have to block deploying to production until the rest of the feature is done to avoid an accidental behavior change for users using the `PRIVATE_STREAM_HISTORY_FOR_SUBSCRIBERS` setting).

eeshangarg commented:
@timabbott: Updated! I refactored some of the duplicated logic that calculates the default value for `history_public_to_subscribers`. :)

timabbott commented:
Can we also use this in `do_change_stream_invite_only`?

eeshangarg commented:
@timabbott: Just updated this. Here are a few thoughts:

* Yes, turns out we can use `get_default_value_for_history_public_to_subscribers` in `do_change_stream_invite_only`. I just missed that, my bad, sorry! :(
* I just added a commit that adds UI for creating a private stream but with public message history. I have never worked on the frontend/JS side of our codebase before, so I am being a bit cautious right now. I don't expect you to merge this commit right away, but I would definitely appreciate a review or an indication as to whether I am on the right track here or not before I go ahead and work on this further! :) I manually tested all of my changes (including stream creation for public, private and private-with-public-history streams).
* I think the failing build is due to the casper flake.

Looking forward to your feedback! Thanks! :)

timabbott commented on static/js/stream_create.js:
```diff
@@ -169,20 +171,34 @@ function create_stream() {
     var principals = get_principals();
 
+    var invite_only = false;
+    var history_public_to_subscribers = true;
```
I would just declare these as `var invite_only` and then set the values in a third `else` clause below.

timabbott commented on static/templates/subscription_creation_form.handlebars:
```diff
@@ -45,6 +45,14 @@
+                        <label class="radio">
+                            <input type="radio" name="privacy" value="invite-only-public-history" />
+                            {{#tr this}}
+                            <b>Private with public message history:</b> must be invited by a member; new members can also see messages sent before they joined; hidden from non-administrator users
```
Maybe "Private with history: must be invited by a member; new members can view complete message history; ..."

timabbott commented:
Overall, the frontend changes look pretty reasonable; posted a few comments.

timabbott commented:
Overall, the frontend changes look pretty reasonable; posted a few comments.

I merged the first commit after fixing the little mypy thing.

timabbott commented:
Scroll up and/or just run the single; the clear output is a bit further up.

timabbott commented:
I did a bit of debugging, and your problem was that you didn't update `notify_subscription_added`.  

Do you understand how this test works?  http://zulip.readthedocs.io/en/latest/subsystems/events-system.html is helpful reading.

timabbott commented:
I merged the first commit after resolving the merge conflict.

eeshangarg commented:
@timabbott: Rebased this! :)

timabbott commented:
Cool, for the last commit, the UI looks pretty good, but I found a few bugs from playing around with it:
* The "Change privacy" modal should always open with the current state selected.
* The "Change privacy" button inside the modal should be consistently red, whether you're changing public->private or private->public.
* If you change a public stream to private, it seems to incorrectly render the private-with-history text here: 
![image](https://user-images.githubusercontent.com/2746074/40136290-52ef403e-58fc-11e8-9643-44b0f3d01af1.png)
* We probably want to rename "Change stream privacy" to "Change stream permissions", because the next thing we're going to do is add @shubham-padia's `is_announce_only` checkbox to this modal.
* We should include the stream name in the template heading here, i.e. "Change stream permissions for #announce":
![image](https://user-images.githubusercontent.com/2746074/40136374-9a6d046e-58fc-11e8-9789-8a046009ebf7.png)

Commit-structure wise, it'd be cleaner to extract the `stream_types.handlebars` extraction (without changes) as its own commit at the beginning; if you'd done that, I would have merged that piece now :).

timabbott commented on static/js/stream_create.js:
```diff
@@ -169,8 +171,23 @@ function create_stream() {
     var stream_name = $.trim($("#create_stream_name").val());
     var description = $.trim($("#create_stream_description").val());
     var is_invite_only = $('#stream_creation_form input[name=privacy]:checked').val() === "invite-only";
+    var is_invite_only_public_history = $('#stream_creation_form input[name=privacy]:checked').val() === "invite-only-public-history";
```
The code would be cleaner and more readable if we did here:

`var private_model = $('#stream_creation_form input[name=privacy]:checked').val()`

And then had the if block below around specific values for this parameter.  

(Both here and in the next file).

eeshangarg commented:
@timabbott: Updated! Ran into a slight bug, but it should be good now! Thanks! :)

eeshangarg commented:
@timabbott: I think the failing build has to do with the casper flake and not my PR :)

timabbott commented on static/js/stream_edit.js:
```diff
@@ -488,8 +506,12 @@ exports.initialize = function () {
+            is_public: !stream.can_make_public,
+            is_private: stream.can_make_public && !stream.history_public_to_subscribers,
+            is_private_with_public_history: (stream.can_make_public &&
+                                             stream.history_public_to_subscribers),
```
This code was wrong; I think it should have been obvious that it was wrong, because if you read it, the things on the RHS didn't line up with the RHS.  I fixed it with the below change:

```
diff --git a/static/js/stream_edit.js b/static/js/stream_edit.js
index 10ce91b..665fde1 100644
--- a/static/js/stream_edit.js
+++ b/static/js/stream_edit.js
@@ -508,9 +508,9 @@ exports.initialize = function () {
         var template_data = {
             stream_id: stream_id,
             stream_name: stream.name,
-            is_public: !stream.can_make_public,
-            is_private: stream.can_make_public && !stream.history_public_to_subscribers,
-            is_private_with_public_history: (stream.can_make_public &&
+            is_public: !stream.invite_only,
+            is_private: stream.invite_only && !stream.history_public_to_subscribers,
+            is_private_with_public_history: (stream.invite_only &&
                                              stream.history_public_to_subscribers),
         };
         var change_privacy_modal = templates.render("subscription_stream_privacy_modal", template_data);
```

Your version seemed to work, but was at best a landmine for future changes.

timabbott commented:
Merged, thanks @eeshangarg!  A few follow-ups:
* `static/templates/subscription_creation_form.handlebars` should be using `stream_types.handlebars` and just passing in the parameters needed to select the correct initial default value of `is_public`.  @shubhamdhama this might be a good project for you, since you've worked with partials recently.
* @rishig was thinking we might want to reorder "private with history" to be second.
* I suspect we want to remove the `can_make_public` and `can_make_private` variables entirely in favor of the new `sub.can_change_subscription_type`; I'm going to try doing that next.
* We can now update the backend to no longer accept values for the `include_history_`  setting of `None` at the API layer; this should let us kill some of that duplicated. code.

timabbott commented:
@novokrest FYI -- you should be able to add the UI for announce-only streams now.
